staging/src/k8s.io/apimachinery/pkg/api/validate/safetraverse/safetraverse.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package safetraverse
+
+// Value takes a pointer to any value, which may or may not be nil,
+// and a function that traverses to another value (a typical use case is to dereference a non-nilable field),
+// and returns the result of the traversal, or returns nil if the given value is nil.
+// This is roughly equivalent to "value ? &fn(value) : nil" in languages that support the ternary operator.
+func Value[V any, T any](value *V, fn func(V) T) *T {
+	if value == nil {
+		return nil
+	}
+	o := fn(*value)
+	return &o
+}
+
+// Ptr takes a pointer to any value, which may or may not be nil,
+// and a function that traverses to another pointer or nilable type (a typical use case is to dereference a nilable field),
+// and returns the result of the traversal, or returns nil if the given value is nil.
+// This is roughly equivalent to "value ? fn(value) : nil" in languages that support the ternary operator.
+func Ptr[V any, T any](value *V, fn func(V) T) T {
+	var result T
+	if value == nil {
+		return result
+	}
+	return fn(*value)
+}
+
+// Lookup takes a map, which may or may not be nil,
+// and a key, and either returns a pointer to the value in the map matching the
+// key, or returns nil.
+func Lookup[K comparable, V any](m map[K]V, k K) *V {
+	if m == nil {
+		return nil
+	}
+	v, ok := m[k]
+	if !ok {
+		return nil
+	}
+	return &v
+}
+
+// Cast takes any value, which may or may not be nil,
+// attempts to cast it to *T and returns *T if the cast is successful, else
+// return nil.
+func Cast[T any](value any) *T {
+	result, _ := value.(*T)
+	return result
+}
+
+// NewListMap creates a ListMap from the given elements and keyer.
+func NewListMap[T any, K comparable](elements []T, keyer func(v T) K) *ListMap[T, K] {
+	if len(elements) == 0 { // optimize for empty maps since this is a common case
+		return nil
+	}
+	keyed := map[K]T{}
+	for _, e := range elements {
+		keyed[keyer(e)] = e
+	}
+	return &ListMap[T, K]{Map: keyed, Keyer: keyer}
+}
+
+// ListMap provides map access to a slice where each element in the slice
+// has a key that can be computed from the element by the provided keyer function.
+type ListMap[T any, K comparable] struct {
+	Map   map[K]T
+	Keyer func(v T) K
+}
+
+// Get returns the element in the ListMap matching the key, or nil if there is no matching element.
+func (c *ListMap[T, K]) Get(key K) *T {
```

Comment:
why is it named `c` ?

Reply:
I think I renamed the type and forgot to rename the receiver arg name to match. Fixing.